types/wlr_xdg_shell.c

```diff
@@ -760,14 +760,19 @@ static void xdg_toplevel_protocol_set_fullscreen(struct wl_client *client,
 	}
 
 	surface->toplevel_state->next.fullscreen = true;
+	surface->toplevel_state->next.fullscreen_output = output;
 
 	struct wlr_xdg_toplevel_set_fullscreen_event event = {
 		.surface = surface,
 		.fullscreen = true,
 		.output = output,
 	};
 
-	wlr_signal_emit_safe(&surface->events.request_fullscreen, &event);
+	if (surface->toplevel_state->added) {
+		wlr_signal_emit_safe(&surface->events.request_fullscreen, &event);
+	} else {
+		wlr_xdg_toplevel_set_fullscreen(surface, true, output);
```

Comment:
I don't agree with this line. The client sends a _request_ to enter fullscreen, the compositor should be able to deny it. `next` contains the state requested by the client, `current` contains the state applied by the server. Instead of adding this line the compositor could check `next` when a surface appears and call `wlr_xdg_toplevel_set_fullscreen` if it wants to.

Reply:
This event never makes it to the compositor in the current state, which is the core problem.

The client sets itself up to the output size and with fullscreen before it does it's first commit (see the referenced issue).
The compositor currently **can't** attach a listener to this event before the commit and therefore can't grant or deny the request.

This just acks everything until the compositor can intercept, and gives the compositor the state the clients wants to have during the first commit.
Should the compositor decide that the client should not have this state to begin with, it can send the appropriate events (size+unset fullscreen) on the first commit.

Reply:
Erm, forgot the important part:
next is cleared at various points because the `wlr_xdg_shell(_v6)` implementation sets it up to the state in `ack_configure`, so we don't get it on first commit if it's just placed on the next state.

Reply:
>This event never makes it to the compositor in the current state, which is the core problem.

No need for an event. Just check `next` in rootston when a surface appears, and send a configure event from there. Don't send configure events without the compositor's consent.